system-addon/content-src/lib/init-store.js

```diff
@@ -76,6 +76,29 @@ const rehydrationMiddleware = store => next => action => {
   return next(action);
 };
 
+// This middleware queues up all the SAVE_SESSION_PERF_DATA actions until it
+// receives the MERGE_STORE_ACTION, which is guaranteed to come from chrome.
+const queueEarlyPerfActionMiddleware = store => next => action => {
+  if (store._ready) {
+    return next(action);
+  }
+
+  if (action.type === MERGE_STORE_ACTION) {
+    next(action);
+    store._ready = true;
+    if (store._queue && store._queue.length) {
+      store._queue.forEach(next);
+    }
+    return null;
+  } else if (action.type === at.SAVE_SESSION_PERF_DATA) {
```

Comment:
This is going to miss on the newly added PAGE_PRERENDERED. I suppose even if we added a `prerendered: true` to NEW_TAB_STATE_REQUEST, we would need to stash that action in `rehydrationMiddleware` to replay the original action with extra data instead of sending a newly created REQUEST.

Reply:
Indeed, would it be possible to queue up all the content-to-main actions and empty them once the main is ready (e.g. "NEW_TAB_INITIAL_STATE")? Not quite sure about that, in particular, given this [test](https://github.com/mozilla/activity-stream/blob/master/system-addon/test/unit/lib/init-store.test.js#L55). Thoughts? @k88hudson

Reply:
After discussing this with @k88hudson, instead of stashing all the early content-to-main actions, we really just wanted to do that for "SAVE_SESSION_PERF_DATA" and "PAGE_PRERENDERED" for now. This middleware will let all other actions (content-to-content, content-to-main, main-to-content) go through immediately.